osu.Game/Overlays/BeatmapListingOverlay.cs

```diff
@@ -195,119 +215,58 @@ private void updateSearch()
 
             loadingLayer.Show();
 
-            getSetsRequest = new SearchBeatmapSetsRequest(
+            beatmapSetPager?.Reset();
+            beatmapSetPager = new BeatmapSetPager(
+                API,
+                rulesets,
                 searchSection.Query.Value,
                 searchSection.Ruleset.Value,
                 searchSection.Category.Value,
                 sortControl.Current.Value,
                 sortControl.SortDirection.Value);
 
-            getSetsRequest.Success += response => Schedule(() => recreatePanels(response));
+            beatmapSetPager.PageFetch += onPageFetch;
+
+            addPageDebounce?.Cancel();
+            addPageDebounce = null;
 
-            API.Queue(getSetsRequest);
+            queueAddPage();
         }
 
-        private void recreatePanels(SearchBeatmapSetsResponse response)
+        private void onPageFetch(List<BeatmapSetInfo> beatmaps)
         {
-            if (response.Total == 0)
+            Schedule(() =>
             {
-                searchSection.BeatmapSet = null;
-                LoadComponentAsync(new NotFoundDrawable(), addContentToPlaceholder);
-                return;
-            }
-
-            var beatmaps = response.BeatmapSets.Select(r => r.ToBeatmapSet(rulesets)).ToList();
+                loadingLayer.Hide();
 
-            var newPanels = new FillFlowContainer<DirectPanel>
-            {
-                RelativeSizeAxes = Axes.X,
-                AutoSizeAxes = Axes.Y,
-                Spacing = new Vector2(10),
-                Alpha = 0,
-                Margin = new MarginPadding { Vertical = 15 },
-                ChildrenEnumerable = beatmaps.Select<BeatmapSetInfo, DirectPanel>(b => new DirectGridPanel(b)
+                if (beatmapSetPager.IsPastFirstPage)
                 {
-                    Anchor = Anchor.TopCentre,
-                    Origin = Anchor.TopCentre,
-                })
-            };
+                    resultsDisplay.AddBeatmaps(beatmaps);
+                }
```

Comment:
Kind of dislike this existing. As I said, if the pager took in a bindable list or was in control of the set of beatmaps, it could clear it itself and with proper event subscriptions this logic becomes cleaner and there's no need to expose `IsPastFirstPage` anymore. I haven't tried it though, if you have and it didn't work then fair enough I suppose.